indexing-service/src/main/java/org/apache/druid/indexing/overlord/GlobalTaskLockbox.java

```diff
@@ -0,0 +1,613 @@
+        task,
+        lockbox -> {
+          lockbox.remove(task);
+          return lockbox.isEmpty();
+        }
+    );
+
+    if (isEmpty) {
+      cleanupLockboxResourceIf(task.getDataSource(), resource -> resource.references.get() <= 0);
+    }
+  }
+
+  @VisibleForTesting
+  Optional<TaskLockbox.TaskLockPosse> getOnlyTaskLockPosseContainingInterval(Task task, Interval interval)
+  {
+    return computeForTask(
+        task,
+        lockbox -> lockbox.getOnlyTaskLockPosseContainingInterval(task, interval)
+    );
+  }
+
+  @VisibleForTesting
+  Set<String> getActiveTasks()
+  {
+    final Set<String> allActiveTasks = new HashSet<>();
+
+    final Set<String> datasourceNames = Set.copyOf(datasourceToLockbox.keySet());
+    for (String datasource : datasourceNames) {
+      allActiveTasks.addAll(
+          computeForDatasource(datasource, TaskLockbox::getActiveTasks)
+      );
+    }
+
+    return allActiveTasks;
+  }
+
+  @VisibleForTesting
+  Map<String, NavigableMap<DateTime, SortedMap<Interval, List<TaskLockbox.TaskLockPosse>>>> getAllLocks()
+  {
+    final Map<String, NavigableMap<DateTime, SortedMap<Interval, List<TaskLockbox.TaskLockPosse>>>>
+        allLocks = new HashMap<>();
+
+    final Set<String> datasourceNames = Set.copyOf(datasourceToLockbox.keySet());
+    for (String datasource : datasourceNames) {
+      final NavigableMap<DateTime, SortedMap<Interval, List<TaskLockbox.TaskLockPosse>>>
+          datasourceLocks = computeForDatasource(datasource, TaskLockbox::getAllLocks);
+      if (!datasourceLocks.isEmpty()) {
+        allLocks.put(datasource, datasourceLocks);
+      }
+    }
+
+    return allLocks;
+  }
+
+  /**
+   * Gets the {@link DatasourceLockboxResource} for the given datasource and
+   * increments the number of references currently in use.
+   * This resource must be closed once it is not needed anymore.
+   */
+  private DatasourceLockboxResource getLockboxResource(String datasource)
+  {
+    return datasourceToLockbox.compute(
+        datasource,
+        (ds, existingResource) -> {
+          final DatasourceLockboxResource resource = Objects.requireNonNullElseGet(
+              existingResource,
+              () -> new DatasourceLockboxResource(
+                  new TaskLockbox(ds, taskStorage, metadataStorageCoordinator)
+              )
+          );
+          resource.acquireReference();
+          return resource;
+        }
+    );
+  }
+
+  /**
+   * Cleans up the lockbox for the given datasource if the {@link DatasourceLockboxResource}
+   * meets the given criteria.
+   */
+  private void cleanupLockboxResourceIf(
+      String dataSource,
+      Predicate<DatasourceLockboxResource> resourcePredicate
+  )
+  {
+    datasourceToLockbox.compute(
+        dataSource,
+        (ds, resource) -> {
+          if (resource != null && resourcePredicate.test(resource)) {
+            // TODO: what if a bad runaway operation is holding the TaskLockbox.giant?
```

Comment:
This is only a concern on shutdown, right? Since at other times, `clear()` will only be called when the refcount is zero, meaning no locks should be held. Just trying to make sure I understand.

Reply:
Yes, it's a concern only on `shutdown()`, which is invoked on loss of leadership.
We would want the leadership change listener to return quickly.

Reply:
True, we do, although we also want the old leader to fully stand down before the new one stands up, to avoid races between the leaders. Ideal thing would be to interrupt other in flight operations, if that's feasible. If not feasible right now, then please update this comment to describe the concern and to not be a TODO comment.

Reply:
Updated and moved the comment to `shutdown()` method as it seems more relevant there.